pkg/storage/etcd3/v3_helper.go

```diff
@@ -0,0 +1,321 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd3
+
+import (
+	"errors"
+	"path"
+	"reflect"
+	"strings"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/conversion"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/storage"
+	"k8s.io/kubernetes/pkg/storage/etcd"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+type v3Helper struct {
+	// etcd interfaces
+	client *clientv3.Client
+
+	codec      runtime.Codec
+	versioner  storage.Versioner
+	pathPrefix string
+}
+
+func newV3Helper(c *clientv3.Client, codec runtime.Codec, prefix string) *v3Helper {
+	return &v3Helper{
+		client:     c,
+		versioner:  etcd.APIObjectVersioner{},
+		codec:      codec,
+		pathPrefix: prefix,
+	}
+}
+
+func (h *v3Helper) Backends(ctx context.Context) []string {
+	if ctx == nil {
+		glog.Errorf("Context is nil")
+		ctx = context.TODO()
+	}
+	resp, err := h.client.MemberList(ctx)
+	if err != nil {
+		glog.Errorf("Error obtaining etcd members list: %q", err)
+		return nil
+	}
+	var mlist []string
+	for _, member := range resp.Members {
+		mlist = append(mlist, member.ClientURLs...)
+	}
+	return mlist
+}
+
+func (h *v3Helper) Codec() runtime.Codec {
+	return h.codec
+}
+
+func (h *v3Helper) Versioner() storage.Versioner {
+	return h.versioner
+}
+
+func (h *v3Helper) Set(ctx context.Context, key string, obj, out runtime.Object, ttl uint64) error {
+	rev, err := extractRevision(h.versioner, obj)
+	if err != nil {
+		return err
+	}
+
+	data, err := runtime.Encode(h.codec, obj)
+	if err != nil {
+		return err
+	}
+	ctx, key = setupRoutine(ctx, key, h.pathPrefix)
+
+	// Note: if rev = 0, it is equivalent to create.
+	txnResp, err := h.client.KV.Txn(ctx).If(
```

Comment:
Hi Tim. If no version existed, it would extract out version of 0, which is equivalent to "create" here.

See https://github.com/kubernetes/kubernetes/blob/586cb9126a45ac2506ed2d08dff6d8095e40e5b8/pkg/storage/etcd/api_object_versioner.go#L71